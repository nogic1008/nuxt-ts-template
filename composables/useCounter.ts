import { computed, useState } from '#app'

export default function () {
  const refCount = useState('count', () => 0)
  const count = computed(() => refCount.value)
  const increment = () => {
    refCount.value++
  }
  const decrement = () => {
    refCount.value--
  }
  return { count, increment, decrement }
}
