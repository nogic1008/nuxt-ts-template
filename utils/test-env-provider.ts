/* eslint-disable no-process-env */
/** Provides process.env for Testing. */
export default class TestEnvProvider<T> {
  /** Stored process.env */
  private readonly storedEnv: NodeJS.ProcessEnv
  private readonly keys: Set<Extract<keyof T, string>>

  constructor(...keys: Extract<keyof T, string>[]) {
    this.storedEnv = { ...process.env }
    this.keys = new Set(keys)
  }

  initialize(obj?: Partial<T>) {
    const env = { ...this.storedEnv }
    this.keys.forEach((k) => delete env[k])
    process.env = { ...env, ...obj }
  }

  restore() {
    process.env = { ...this.storedEnv }
  }
}
