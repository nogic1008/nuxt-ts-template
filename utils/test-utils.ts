/** Generate [0-9a-z]{length} string. */
export const generateRandomString = (length: number) =>
  [...Array(Number.isInteger(length) ? length : 12)]
    .map(() => (~~(Math.random() * 36)).toString(36))
    .join('')
