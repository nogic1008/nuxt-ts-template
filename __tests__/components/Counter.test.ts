import { beforeEach, describe, expect, jest, test } from '@jest/globals'
import { ref } from '@nuxtjs/composition-api'
import { mount } from '@vue/test-utils'

import { createI18n, createVue } from '~/__tests__/utils'
import CounterComponent from '~/components/Counter.vue'
import useCounter from '~/composables/useCounter'

jest.mock('~/composables/useCounter')

const localVue = createVue()

describe('components/Counter.vue', () => {
  const increment = jest.fn()
  const decrement = jest.fn()
  const i18n = createI18n()
  beforeEach(() => {
    jest
      .mocked(useCounter)
      .mockReturnValue({ count: ref(0), increment, decrement })
    increment.mockClear()
    decrement.mockClear()
  })

  describe.each(['en', 'ja'])('snapshot test (%s)', (locale: string) => {
    test('renders correctly', () => {
      // Arrange
      const i18n = createI18n(locale)

      // Act
      const wrapper = mount(CounterComponent, { localVue, i18n })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  test.each([1, 2, 10])(
    'click plus button calls useCounter().increment()',
    async (pressCount: number) => {
      // Arrange
      const wrapper = mount(CounterComponent, { localVue, i18n })

      // Act
      const button = wrapper.find('button.plus')
      for (let i = 0; i < pressCount; i++) {
        await button.trigger('click')
      }

      // Assert
      expect(increment).toBeCalledTimes(pressCount)
      expect(decrement).not.toBeCalled()
    }
  )

  test.each([1, 2, 10])(
    'click minus button calls useCounter().decrement()',
    async (pressCount: number) => {
      // Arrange
      const wrapper = mount(CounterComponent, { localVue, i18n })

      // Act
      const button = wrapper.find('button.minus')
      for (let i = 0; i < pressCount; i++) {
        await button.trigger('click')
      }

      // Assert
      expect(decrement).toBeCalledTimes(pressCount)
      expect(increment).not.toBeCalled()
    }
  )
})
