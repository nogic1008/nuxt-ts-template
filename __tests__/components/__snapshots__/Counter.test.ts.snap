// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Counter.vue snapshot test (en) renders correctly 1`] = `
<div
  class="column"
>
  <div
    class="card"
  >
    <header
      class="card-header"
    >
      <p
        class="card-header-title has-text-grey"
      >
        
        Store Sample
      
      </p>
    </header>
     
    <div
      class="card-content"
    >
      <div
        class="content has-text-centered"
      >
        <span
          class="icon is-large has-text-primary"
        >
          <i
            class="mdi mdi-counter mdi-48px"
          />
        </span>
      </div>
    </div>
     
    <footer
      class="card-footer"
    >
      <div
        class="card-footer-item"
      >
        <span>
          <button
            class="plus button"
            type="button"
          >
            <span
              class="button-wrapper"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-plus mdi-24px"
                />
              </span>
              <!---->
              <!---->
            </span>
          </button>
           
          <span
            class="subtitle"
          >
            0
          </span>
           
          <button
            class="minus button"
            type="button"
          >
            <span
              class="button-wrapper"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-minus mdi-24px"
                />
              </span>
              <!---->
              <!---->
            </span>
          </button>
        </span>
      </div>
    </footer>
  </div>
</div>
`;

exports[`components/Counter.vue snapshot test (ja) renders correctly 1`] = `
<div
  class="column"
>
  <div
    class="card"
  >
    <header
      class="card-header"
    >
      <p
        class="card-header-title has-text-grey"
      >
        
        Store サンプル
      
      </p>
    </header>
     
    <div
      class="card-content"
    >
      <div
        class="content has-text-centered"
      >
        <span
          class="icon is-large has-text-primary"
        >
          <i
            class="mdi mdi-counter mdi-48px"
          />
        </span>
      </div>
    </div>
     
    <footer
      class="card-footer"
    >
      <div
        class="card-footer-item"
      >
        <span>
          <button
            class="plus button"
            type="button"
          >
            <span
              class="button-wrapper"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-plus mdi-24px"
                />
              </span>
              <!---->
              <!---->
            </span>
          </button>
           
          <span
            class="subtitle"
          >
            0
          </span>
           
          <button
            class="minus button"
            type="button"
          >
            <span
              class="button-wrapper"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-minus mdi-24px"
                />
              </span>
              <!---->
              <!---->
            </span>
          </button>
        </span>
      </div>
    </footer>
  </div>
</div>
`;
